### PR TITLE
[CFE-381]: Feature - bonded changes transaction

### PR DESCRIPTION
**Pull request by Poafs1**
Screenshot 2567-03-19 at 15 49 28
Screenshot 2567-03-19 at 15 49 36

### CODE REVIEW

File: src/lib/services/validator.ts [hunk 1]
```diff
@@ -214,13 +214,18 @@ export const getValidatorUptime = async (
...
+
 const zValidatorDelegationRelatedTxsResponseItem = z
   .object({
     tx_hash: z.string(),
     height: z.number().positive(),
     tokens: zCoin.array(),
     timestamp: zUtcDate,
-    validator_address: zValidatorAddr,
+    messages: z.array(zValidatorDelegationRelatedTxsResponseMessage),
+    sender: zValidatorAddr,
```
Comment (songwongtp) [inline src/lib/services/validator.ts:228]: zBechAddr

**Review by songwongtp — APPROVED** [review]

File: src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsTable.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import { RelatedTransactionsTableRow } from "./RelatedTransactionsTableRow";
+
+interface RelatedTransactionTableProps {
+  delegationRelatedTxs: Option<ValidatorDelegationRelatedTxsResponseItem[]>;
+  isLoading: boolean;
+  assetInfos: Option<AssetInfos>;
+}
+
+export const RelatedTransactionTable = ({
+  delegationRelatedTxs = [],
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsTable.tsx:16]: remove default `[]`

File: src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsTable.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  isLoading,
+  assetInfos,
+}: RelatedTransactionTableProps) => {
+  if (isLoading) return <Loading />;
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsTable.tsx:20]:
```
if (!delegationRelatedTxs) return <ErrorFetching dataName="delegation related txs" />
```

File: src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsTableRow.tsx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  templateColumns: string;
+  assetInfos: Option<AssetInfos>;
+}
+
+export const RelatedTransactionsTableRow = ({
+  delegationRelatedTx,
+  templateColumns,
+  assetInfos,
+}: RelatedTransactionsTableRowProps) => {
+  return (
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsTableRow.tsx:22]: implicite return

File: src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsTable.tsx
Comment (songwongtp) [inline src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsTable.tsx:1]: let's make this file `index.tsx` of this directory

File: src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsMobileCard.tsx
Comment (songwongtp) [inline src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsMobileCard.tsx:1]: MobileCard is called inside of the Table component in this case is `RelatedTransactionsTable`. Please see the other table as an example

File: src/lib/pages/validator-details/components/bonded-token-changes/index.tsx [hunk 5]
```diff
@@ -15,13 +23,82 @@ export const BondedTokenChanges = ({
...
+              id={tableHeaderId}
+              helperText="Shows transactions relevant to changes in delegated tokens, excluding any token reduction due to slashing."
+            />
+            <RelatedTransactionTable
+              delegationRelatedTxs={data?.items}
+              isLoading={isLoading}
+              assetInfos={assetInfos}
+            />
+          </>
+        )}
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/bonded-token-changes/index.tsx:84]:
```
{
  !isMobile && (
    <TableTitle
              title="Delegation-Related Transactions"
              count={data?.total ?? 0}
              id={tableHeaderId}
              helperText="Shows transactions relevant to changes in delegated tokens, excluding any token reduction due to slashing."
            />
  )
  <RelatedTransactionTable
              delegationRelatedTxs={data?.items}
              isLoading={isLoading}
              assetInfos={assetInfos}
   />
}
```

**Review by songwongtp — CHANGES_REQUESTED** [review]

File: src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsBondedTokenChanges.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+import type { AssetInfos, Coin, Option, USD } from "lib/types";
+import {
+  coinToTokenWithValue,
+  formatPrice,
+  formatUTokenWithPrecision,
+} from "lib/utils";
+
+interface RelatedTransactionsBondedTokenChangesProps {
+  txHash: string;
+  token: Coin;
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsBondedTokenChanges.tsx:15]: `coin: Coin`

File: src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsBondedTokenChanges.tsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+export const RelatedTransactionsBondedTokenChanges = ({
+  txHash,
+  token,
+  assetInfos,
+}: RelatedTransactionsBondedTokenChangesProps) => {
+  const tokenWithValue = coinToTokenWithValue(
+    token?.denom,
+    token?.amount,
+    assetInfos
+  );
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsBondedTokenChanges.tsx:28]:
```
const token = coinToTokenWithValue(
    coin.denom,
    coin.amount,
    assetInfos
  );
```

File: src/lib/pages/validator-details/components/tables/related-transactions/index.tsx [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+  if (isLoading) return <Loading />;
+  if (!delegationRelatedTxs)
+    return <ErrorFetching dataName="delegation related txs" />;
+
+  const templateColumns = "max(180px) max(180px) max(180px) 1fr max(280px)";
+
+  return isMobile ? (
+    <MobileTableContainer>
+      {delegationRelatedTxs?.map((delegationRelatedTx) => (
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/tables/related-transactions/index.tsx:35]: remove `?`

File: src/lib/services/validator.ts [hunk 9]
```diff
@@ -220,9 +225,20 @@ const zValidatorDelegationRelatedTxsResponseItem = z
...
+        type: z.string(),
+      })
+    ),
+    sender: zBechAddr,
   })
-  .transform(snakeToCamel);
+  .transform((val) => ({
+    ...snakeToCamel(val),
+    txHash: parseTxHash(val.tx_hash),
+  }));
```
Comment (songwongtp) [inline src/lib/services/validator.ts:238]: this is already correct, but you can also do
```
z.object({
  tx_hash: z.string().transform(parseTxHash),
  ...
}).transform(snakeToCamel);
```

**Review by songwongtp — CHANGES_REQUESTED** [review]

File: src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsBondedTokenChanges.tsx [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+  const formattedAmount = formatUTokenWithPrecision(
+    tokenWithValue.amount,
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsBondedTokenChanges.tsx:31]: let's do
```
const formattedAmount = `{isPositiveAmount ? "+":"-"}{formatUTokenWithPrecision(tokenWithValue.amount.abs(), tokenWithValue.precision ?? 0, false, 2)}`
```

File: src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsBondedTokenChanges.tsx [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+      alignItems="center"
+    >
+      <Box textAlign={{ base: "left", md: "right" }}>
+        <Text>
+          <Text
+            as="span"
+            fontWeight={700}
+            color={isPositiveAmount ? "success.main" : "error.main"}
+          >
+            {isPositiveAmount ? `+${formattedAmount}` : formattedAmount}
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsBondedTokenChanges.tsx:55]: `{formattedAmount}` as we already handle +/- above

File: src/lib/utils/formatter/token.ts [hunk 12]
```diff
@@ -68,18 +68,23 @@ export const formatUTokenWithPrecision = (
   amount: U<Token<BigSource>>,
   precision: number,
   isSuffix = true,
-  decimalPoints?: number
+  decimalPoints?: number,
+  allowNegativeValue = false
```
Comment (songwongtp) [inline src/lib/utils/formatter/token.ts:72]: remove

File: src/lib/utils/formatter/token.ts [hunk 13]
```diff
@@ -68,18 +68,23 @@ export const formatUTokenWithPrecision = (
...
-  const lowestThreshold = big(10).pow(-(decimalPoints ?? precision));
-  if (!token.eq(0) && token.lt(lowestThreshold)) {
-    return `<${lowestThreshold.toFixed()}`;
+  if (!allowNegativeValue) {
+    const lowestThreshold = big(10).pow(-(decimalPoints ?? precision));
+
+    if (!token.eq(0) && token.lt(lowestThreshold)) {
+      return `<${lowestThreshold.toFixed()}`;
+    }
   }
```
Comment (songwongtp) [inline src/lib/utils/formatter/token.ts:88]: revert back

**Review by songwongtp — CHANGES_REQUESTED** [review]

File: src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsBondedTokenChanges.tsx [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+          </Text>{" "}
+          {tokenWithValue.symbol}
+        </Text>
+        <Text variant="body3" color="text.dark">
+          ({formatPrice(tokenWithValue.value?.abs() as USD<BigSource>)})
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsBondedTokenChanges.tsx:60]: `tokenWithValue.value ? formatPrice(tokenWithValue.value.abs() as USD<BigSource> : "-"`

**Review by songwongtp — CHANGES_REQUESTED** [review]

**Review by songwongtp — APPROVED** [review]

File: src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsBondedTokenChanges.tsx [hunk 15]
```diff
@@ -43,7 +44,7 @@ export const RelatedTransactionsBondedTokenChanges = ({
           >
             {formattedAmount}
           </Text>{" "}
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/tables/related-transactions/RelatedTransactionsBondedTokenChanges.tsx:46]: remove `{" "}`

File: src/lib/pages/validator-details/components/tables/related-transactions/index.tsx [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+  isLoading: boolean;
+  assetInfos: Option<AssetInfos>;
+  onRowSelect: (txHash: string) => void;
+}
+
+export const RelatedTransactionTable = ({
+  delegationRelatedTxs,
+  isLoading,
+  assetInfos,
+  onRowSelect,
```
Comment (evilpeach) [inline src/lib/pages/validator-details/components/tables/related-transactions/index.tsx:23]: It's no need pass from parent. Let's create navigate function here.

**Review by evilpeach — APPROVED** [review]

**Pull request merged**